### PR TITLE
+ Create: assign orders to drivers logic

### PR DESCRIPTION
**Pull request by Parle648**
## Describe your changes

- In this PR, I moved the logic for assigning orders to drivers from the front-end to the back-end.

## Issue ticket code (and/or) and link

- [Link to JIRA ticket](https://codecraftersintership.atlassian.net/browse/SCRUM-118)

### **General**

- [x] types for input and output params
- [x] no `any`, should be also added to eslint rules
- [x] try/catch for any async function
- [x] no **magic** [numbers](/7d77574ee4bc4e339a68066762e887cb)
- [x] compare only with constants not with strings (instead of user === ‘admin’, better user === roles.admin)
- [x] no ternary operator inside ternary operator (bad example: question ? first: second ? third: fourth)
- [x] avoid similar types with duplicating by generics
- [x] no **console.log** in pr
- [x] .env file should be in .gitignore
- [x] delete commented code if it's not part of documentation
- [x] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file
- [x] **constants** and **function** names should be lowercase.
- [x] no dates format in the code (like "dd/MM/yyyy”), move it in global constant variable
- [x] import rules. imports should come in a specific order: node modules, absolute path, relative path
- [x] strings should be in the constant variable. Example: instead of ‘15 3 \* \* _’, const EACH_DAY_15h_15min = ‘15 3 _ \* \*’

### Backend

- [x] swagger for each endpoint. add in the documentation, different types of responses.
      on a successful response (2xx), on unsuccessful response (4xx, 5xx)
- [x] less requests to db, all data should be taken with one query
- [x] **public** in method use only is use function externally
- [x] use ConfigService instead of **process.env**
- [x] use @`@index` decorator for frequently requested data
- [x] use transactions if there is a call chain that mutates data in different tables
- [x] add a set of rules for [nestjs](https://github.com/darraghoriordan/eslint-plugin-nestjs-typed) to the .eslint config file. It helps to prevent bugs and confusing moments
- [x] use [REST API Naming Convention](https://medium.com/@nadinCodeHat/rest-api-naming-conventions-and-best-practices-1c4e781eb6a5) for endpoints

  P.S: more information about RESTful API design by [Microsof](https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design)t

- [x] use ‘**UUIDs**` for primary key, ** to prevent Enumeration Attacks**

### CODE REVIEW

**Review by AlexeyBezrebriy — APPROVED** [review]
Looks good

**Review by oyakovytskyi — APPROVED** [review]

File: src/modules/assign-orders/assign-orders.controller.ts
Comment (lyudad) [inline src/modules/assign-orders/assign-orders.controller.ts:1]: I dont think we need to create separate module for it. move it to orders module

File: src/modules/assign-orders/assign-orders.controller.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  constructor(private readonly assignOrdersService: AssignOrdersService) {}
+
+  @Get()
+  @UseGuards(RolesGuard)
+  @SetMetadata('roles', [Roles.DISPATCHER])
+  @ApiOperation({ summary: 'Get assigned orders for the selected drivers' })
+  @ApiResponse({ status: 200, example: { value: [{ driver: User, orders: [Order] }], notAssignedOrders: [Order] } })
+  @ApiResponse({ status: 500, type: FailedResponse })
+  async createRoutes(@Query() { driversIds, ordersIds }: { driversIds: string; ordersIds: string }): Promise<AssignedOrdersResponse> {
+    return this.assignOrdersService.getAndAssign(<number[]>JSON.parse(driversIds), <number[]>JSON.parse(ordersIds));
```
Comment (lyudad) [inline src/modules/assign-orders/assign-orders.controller.ts]: you can write parser as decorator

File: src/modules/assign-orders/assign-orders.service.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+export class AssignOrdersService {
+  constructor(
+    @InjectRepository(User)
+    private readonly userRepo: Repository<User>,
+    @InjectRepository(Order)
+    private readonly orderRepo: Repository<Order>,
+  ) {}
+
+  private assignOrdersToDrivers(drivers: User[], orders: Order[]): AssignedOrdersResponse {
+    orders.sort((a, b) => new Date(a.collection_time_start).getTime() - new Date(b.collection_time_start).getTime());
```
Comment (lyudad) [inline src/modules/assign-orders/assign-orders.service.ts]: empty line below

File: src/modules/assign-orders/assign-orders.service.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    const assignments: { driver: User; orders: Order[] }[] = drivers.map((driver) => ({
+      driver,
+      orders: [],
+    }));
```
Comment (lyudad) [inline src/modules/assign-orders/assign-orders.service.ts]: empty line below

File: src/modules/assign-orders/assign-orders.controller.ts (hunk 1 above)
Comment (Parle648) [reply src/modules/assign-orders/assign-orders.controller.ts]: create parser (it works without @Query)

File: src/modules/assign-orders/assign-orders.controller.ts
Comment (Parle648) [reply src/modules/assign-orders/assign-orders.controller.ts:1]: moved it to orders module

File: src/modules/assign-orders/assign-orders.service.ts (hunk 2 above)
Comment (Parle648) [reply src/modules/assign-orders/assign-orders.service.ts]: done

File: src/modules/assign-orders/assign-orders.service.ts (hunk 3 above)
Comment (Parle648) [reply src/modules/assign-orders/assign-orders.service.ts]: done

**Review by oyakovytskyi — APPROVED** [review]

**Review by lyudad — APPROVED** [review]

**Pull request merged**